File: android/app/src/main/java/com/dbtutorialone/BiometricAuthModule.java
package com.dbtutorialone;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.biometric.BiometricManager;
import androidx.fragment.app.FragmentActivity;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

public class BiometricAuthModule extends ReactContextBaseJavaModule {
    private String titlePrompt = "";
    private String subTitlePrompt = "";
    private String cancelPrompt = "";

    public BiometricAuthModule(ReactApplicationContext reactAppContext) {
        super(reactAppContext);
    }

    @NonNull
    @Override
    public String getName() {
        return "BiometricAuthModule";
    }

    @ReactMethod
    public void isBiometricSupported(Callback callback) {
        FragmentActivity activity = (FragmentActivity) getCurrentActivity();
        if (activity == null) {
            callback.invoke("Activity is null");
            return;
        }

        BiometricManager biometricManager = BiometricManager.from(activity);
        int biometricAuthResult = biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK);
        String result;
        switch (biometricAuthResult) {
            case BiometricManager.BIOMETRIC_SUCCESS:
                result = ErrorCode.BIOMETRIC_SUCCESS;
                break;
            case BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE:
                result = ErrorCode.BIOMETRIC_ERROR_NO_HARDWARE;
                break;
            case BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE:
                result = ErrorCode.BIOMETRIC_ERROR_HW_UNAVAILABLE;
                break;
            case BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED:
                result = ErrorCode.BIOMETRIC_ERROR_NONE_ENROLLED;
                break;
            case BiometricManager.BIOMETRIC_ERROR_UNSUPPORTED:
                result = ErrorCode.BIOMETRIC_ERROR_UNSUPPORTED;
                break;
            default:
                result = ErrorCode.BIOMETRIC_STATUS_UNKNOWN;
        }
        callback.invoke(result);
    }

    @ReactMethod
    public void authenticationPrompt(String titleArg, String subTitleArg, String cancelArg) {
        titlePrompt = titleArg;
        subTitlePrompt = subTitleArg;
        cancelPrompt = cancelArg;
    }

    @ReactMethod
    public void authenticate(final Callback callback) {
        final FragmentActivity activity = (FragmentActivity) getCurrentActivity();
        if (activity == null) {
            callback.invoke("Activity is null");
            return;
        }

        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                BiometricAuth biometricAuth = new BiometricAuth(activity);
                biometricAuth.authenticate(new BiometricAuth.BiometricCallback() {
                    @Override
                    public void onSuccess() {
                        callback.invoke(null, "Authentication successful");
                    }

                    @Override
                    public void onError(int errorCode, @NonNull String errorString) {
                        callback.invoke("Authentication Error: " + errorString);
                    }

                    @Override
                    public void onAuthFailed() {
                        callback.invoke("Authentication failed");
                    }

                    @NonNull
                    @Override
                    public String getTitle() {
                        return titlePrompt;
                    }

                    @NonNull
                    @Override
                    public String getSubTitle() {
                        return subTitlePrompt;
                    }

                    @NonNull
                    @Override
                    public String getCancel() {
                        return cancelPrompt;
                    }
                });
            }
        });
    }

    public interface ErrorCode {
        String BIOMETRIC_SUCCESS = "BIOMETRIC_SUCCESS";
        String BIOMETRIC_STATUS_UNKNOWN = "BIOMETRIC_STATUS_UNKNOWN";
        String BIOMETRIC_ERROR_UNSUPPORTED = "BIOMETRIC_ERROR_UNSUPPORTED";
        String BIOMETRIC_ERROR_HW_UNAVAILABLE = "BIOMETRIC_ERROR_HW_UNAVAILABLE";
        String BIOMETRIC_ERROR_NONE_ENROLLED = "BIOMETRIC_ERROR_NONE_ENROLLED";
        String BIOMETRIC_ERROR_NO_HARDWARE = "BIOMETRIC_ERROR_NO_HARDWARE";
        String AUTHENTICATION_FAILED = "AUTHENTICATION_FAILED";
        String AUTHENTICATION_ERROR = "AUTHENTICATION_ERROR";
    }
}

File: android/app/src/main/java/com/dbtutorialone/BiometricAuthModule.kt
package com.dbtutorialone

import android.os.Handler
import android.os.Looper
import android.widget.Toast
import androidx.biometric.BiometricManager
import androidx.fragment.app.FragmentActivity
import com.facebook.react.bridge.Callback
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod

class BiometricAuthModule(reactAppContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactAppContext) {

    var titlePrompt: String = ""

    override fun getName(): String {
        return "BiometricAuthModule"
    }

    @ReactMethod
    fun isBiometricSupported(callback: Callback) {
        val context = currentActivity as FragmentActivity

        val biometricManager = BiometricManager.from(context)
        val result =
            when (biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK)) {

                BiometricManager.BIOMETRIC_SUCCESS -> {
                    // showMessage(ErrorCode.BIOMETRIC_SUCCESS.message)
                    true
                }

                BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE -> {
                    showMessage(ErrorCode.BIOMETRIC_ERROR_NO_HARDWARE.message)
                    false
                }

                BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE -> {
                    showMessage(ErrorCode.BIOMETRIC_ERROR_HW_UNAVAILABLE.message)
                    false
                }

                BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED -> {
                    showMessage(ErrorCode.BIOMETRIC_ERROR_NONE_ENROLLED.message)
                    false
                }

                BiometricManager.BIOMETRIC_ERROR_UNSUPPORTED -> {
                    showMessage(ErrorCode.BIOMETRIC_ERROR_UNSUPPORTED.message)
                    false
                }

                else -> {
                    showMessage(ErrorCode.BIOMETRIC_STATUS_UNKNOWN.message)
                    false
                }
            }

        callback.invoke(result)
    }

    @ReactMethod
    fun promptTitle(title: String) {

        titlePrompt = title;
    }

    @ReactMethod
    fun authenticate(callback: Callback) {
        val activity = currentActivity as FragmentActivity

        Handler(Looper.getMainLooper()).post {
            val biometricAuth = BiometricAuth(activity)
            biometricAuth.authenticate(object : BiometricAuth.BiometricCallback {

                override fun onSuccess() {
                    callback.invoke(null, ErrorCode.BIOMETRIC_SUCCESS)
                }

                override fun onError() {
                    callback.invoke(ErrorCode.AUTHENTICATION_ERROR)
                }

                override val title: String
                    get() = titlePrompt

            })

        }
    }


    @ReactMethod
    private fun showMessage(message: String) {
        Toast.makeText(reactApplicationContext, message, Toast.LENGTH_LONG).show()
    }

    private enum class ErrorCode(val message: String) {
        BIOMETRIC_SUCCESS("Biometric Success"),
        BIOMETRIC_STATUS_UNKNOWN("Biometric status Unknown"),
        BIOMETRIC_ERROR_UNSUPPORTED("Biometric Error Unsupported"),
        BIOMETRIC_ERROR_HW_UNAVAILABLE("Biometric Error Hardware Unavailable"),
        BIOMETRIC_ERROR_NONE_ENROLLED("Biometric Error None Enrolled"),
        BIOMETRIC_ERROR_NO_HARDWARE("Biometric Error No Hardware"),
        AUTHENTICATION_FAILED("Biometric is registered but not authorised with Current User."),
        AUTHENTICATION_ERROR("Authentication Error");

    }

}

